### PR TITLE
ZoneHVAC:PackagedTerminalAirConditioner with autosized steam coil terminates with fatal error

### LINKED ISSUE #7480: ZoneHVAC:PackagedTerminalAirConditioner with autosized steam coil terminates
Issue overview
--------------
A user used steam coils in the ZoneHVAC:PackagedTerminalAirConditioner and autosized the Maximum Steam Flow Rate field. The program tried to size a water coil instead of a steam coil and terminated the simulation with a fatal error.

Workaround: Hard size the Maximum Steam Flow Rate prior to V9.2 (e.g., 0.01 m3/s).
Helpdesk ticket: 15260

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
This fixes #7480. When a steam coil is used as the heating coil in ZoneHVAC:PackagedTerminalAirConditioner and the steam heating coil Max Steam Flow Rate is autosized, the simulation terminates with a fatal error that a water coil could not be found.

Work around: Manually size the steam coil Max Steam Flow Rate.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

File: src/EnergyPlus/BoilerSteam.cc [hunk 1]
```diff
@@ -394,7 +394,7 @@ namespace BoilerSteam {
                 ErrorsFound = true;
             }
 
-            if (rNumericArgs(5) == 0.0) {
+            if (rNumericArgs(5) < 0.0) {
```
Comment (rraustad) [inline src/EnergyPlus/BoilerSteam.cc:397]: IDD shows allowed boiler min flow rate = 0, defect file halted here.

File: src/EnergyPlus/PackagedTerminalHeatPump.cc [hunk 2]
```diff
@@ -4243,17 +4243,30 @@ namespace PackagedTerminalHeatPump {
 
             //   set fluid-side hardware limits
             if (PTUnit(PTUnitNum).HeatCoilFluidInletNode > 0) {
-                // If water coil max water flow rate is autosized, simulate once in order to mine max water flow rate
+                // If coil max fluid flow rate is autosized, simulate once in order to mine max flow rate
```
Comment (rraustad) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc:4246]: Differentiate water and steam coils to avoid crash.

**Comment by rraustad** [comment]
Prior to this fix the PTAC would mine the water coil flow rate for both the water and steam coils. Of course this would cause a fatal termination when an autosized steam coil was used as the heating coil.

```
   **  Fatal  ** SimulateWaterCoilComponents: Invalid CompIndex passed=1, Number of Water Coils=0, Coil name=PTAC HW HTG COIL
```

Once this was corrected, the simulation terminated with an invalid min fluid flow rate in Boiler:Steam. Since a 0 is allowed for this field, a minor change to that warning message was made (warn if < 0, not == 0).

```
Boiler:Steam,
  N5 , \field Minimum Part Load Ratio
       \minimum 0.0
```

After these changes, the simulation completes and the steam coils size as expected.

```
  ************* EnergyPlus Completed Successfully-- 44 Warning; 0 Severe Errors; Elapsed Time=00hr 01min 14.60sec
```

Partial sizing details:

```
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL, Maximum Steam Flow Rate [m3/s], 1.39295E-002
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL 11, Maximum Steam Flow Rate [m3/s], 2.22059E-003
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL 4, Maximum Steam Flow Rate [m3/s], 2.22045E-003
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL 12, Maximum Steam Flow Rate [m3/s], 2.80249E-003
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL 8, Maximum Steam Flow Rate [m3/s], 2.80948E-003
 Component Sizing Information, Coil:Heating:Steam, PTAC HW HTG COIL 1, Maximum Steam Flow Rate [m3/s], 2.21617E-003
```

**Comment by rraustad** [comment]
@Myoldmopar a simple change to correct a HelpDesk issue.

File: src/EnergyPlus/BoilerSteam.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/BoilerSteam.cc:397]: Makes sense that you should only be warning if it's less than zero.

File: src/EnergyPlus/PackagedTerminalHeatPump.cc [hunk 3]
```diff
@@ -4243,17 +4243,30 @@ namespace PackagedTerminalHeatPump {
...
+                        SimulateWaterCoilComponents(PTUnit(PTUnitNum).ACHeatCoilName, FirstHVACIteration, PTUnit(PTUnitNum).ACHeatCoilIndex);
+                        CoilMaxVolFlowRate = GetCoilMaxWaterFlowRate("Coil:Heating:Water", PTUnit(PTUnitNum).ACHeatCoilName, ErrorsFound);
+                        if (CoilMaxVolFlowRate != AutoSize) {
+                            rho = GetDensityGlycol(PlantLoop(PTUnit(PTUnitNum).HeatCoilLoopNum).FluidName,
+                                                   DataGlobals::HWInitConvTemp,
+                                                   PlantLoop(PTUnit(PTUnitNum).HeatCoilLoopNum).FluidIndex,
+                                                   RoutineNameSpace);
+                            PTUnit(PTUnitNum).MaxHeatCoilFluidFlow = CoilMaxVolFlowRate * rho;
+                        }
+                    } else if (PTUnit(PTUnitNum).ACHeatCoilType_Num == Coil_HeatingSteam) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc:4258]: Yep, splitting out heating coil makes good sense.

File: src/EnergyPlus/PackagedTerminalHeatPump.cc [hunk 4]
```diff
@@ -4243,17 +4243,30 @@ namespace PackagedTerminalHeatPump {
...
+                        SimulateSteamCoilComponents(PTUnit(PTUnitNum).ACHeatCoilName,
```
Comment (Myoldmopar) [inline src/EnergyPlus/PackagedTerminalHeatPump.cc:4259]: Does calling Simulate here have any side effects?  It seems like you don't really need to simulate, just make sure input is processed and mine out the max steam flow rate, but I understand that calling Simulate is generally the safest way to ensure everything has been processed.

File: tst/EnergyPlus/unit/PackagedTerminalHeatPump.unit.cc [hunk 5]
```diff
@@ -538,6 +538,8 @@ TEST_F(EnergyPlusFixture, PackagedTerminalHP_VSCoils_Sizing)
     Real64 OnOffAirFlowRatio(1.0); // ratio of compressor ON airflow to average airflow over timestep
     Real64 ZoneLoad(0.0);          // cooling or heating needed by zone [watts]
 
+    // Also set BeginEnvrnFlag so code is tested for coil initialization and does not crash
+    DataGlobals::BeginEnvrnFlag = true;
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/PackagedTerminalHeatPump.unit.cc:542]: A reasonable change.

**Review by Myoldmopar — APPROVED** [review]

File: src/EnergyPlus/PackagedTerminalHeatPump.cc (hunk 4 above)
Comment (rraustad) [reply src/EnergyPlus/PackagedTerminalHeatPump.cc:4259]: Agreed, might be a better way to get the info, it's just what was here. So it's open to future modification.

**Comment by Myoldmopar** [comment]
This was showing no diffs before all the table diffs came in from the other branch.  This is good to go.

**Pull request merged**